mlrun/artifacts/model.py

```diff
@@ -399,22 +426,71 @@ def before_log(self):
             self.metadata.labels = self.metadata.labels or {}
             self.metadata.labels["framework"] = self.spec.framework
 
-    def upload(self):
-
-        target_model_path = path.join(self.spec.target_path, self.spec.model_file)
+    def upload(self, artifact_path: str = None):
+        """
+        internal, upload to target store
+        :param artifact_path: required only for when generating target_path from artifact hash
+        """
+        # if mlrun.mlconf.should_generate_target_path_from_artifact_hash() outputs True and the user
+        # didn't pass target_path explicitly, then target_path will be calculated right before uploading the artifact
+        # using `resolve_<body/file>_target_hash_path`
+        target_model_path = (
+            path.join(self.spec.target_path, self.spec.model_file)
+            if self.spec.target_path
+            else self.spec.target_path
+        )
         body = self.spec.get_body()
         if body:
-            self._upload_body(body, target=target_model_path)
+            if not target_model_path:
+                (
+                    self.metadata.hash,
+                    target_model_path,
+                ) = self.resolve_body_target_hash_path(
+                    body=body, artifact_path=artifact_path
+                )
+            self._upload_body(
+                body, target=target_model_path, artifact_path=artifact_path
+            )
         else:
             src_model_path = _get_src_path(self, self.spec.model_file)
             if not path.isfile(src_model_path):
                 raise ValueError(f"model file {src_model_path} not found")
-            self._upload_file(src_model_path, target=target_model_path)
 
-        upload_extra_data(self, self.spec.extra_data)
+            if not target_model_path:
+                (
+                    self.metadata.hash,
+                    target_model_path,
+                ) = self.resolve_file_target_hash_path(
+                    source_path=src_model_path, artifact_path=artifact_path
+                )
+
+            self._upload_file(
+                src_model_path,
+                target_path=target_model_path,
+                artifact_path=artifact_path,
+            )
 
-        spec_path = path.join(self.spec.target_path, model_spec_filename)
-        store_manager.object(url=spec_path).put(self.to_yaml())
+        upload_extra_data(
+            artifact=self, extra_data=self.spec.extra_data, artifact_path=artifact_path
+        )
+        _, spec_target_path = self.resolve_body_target_hash_path(
```

Comment:
So now you're always uploading it to a file whose name is the hash of the model spec? If not using hash for names you should still stick to the "old" implementation which uses `model_spec_filename`